<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>8 Pipeline Outputs | SPEAQeasy</title>
  <meta name="description" content="8 Pipeline Outputs | SPEAQeasy" />
  <meta name="generator" content="bookdown 0.20 and GitBook 2.6.7" />

  <meta property="og:title" content="8 Pipeline Outputs | SPEAQeasy" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="8 Pipeline Outputs | SPEAQeasy" />
  
  
  

<meta name="author" content="Nicholas J. Eagles" />



  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="manifest.html"/>
<link rel="next" href="configuration.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<script src="libs/accessible-code-block-0.0.1/empty-anchor.js"></script>



</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Overview</a></li>
<li class="chapter" data-level="2" data-path="quick-start.html"><a href="quick-start.html"><i class="fa fa-check"></i><b>2</b> Quick Start</a><ul>
<li class="chapter" data-level="2.1" data-path="quick-start.html"><a href="quick-start.html#setup"><i class="fa fa-check"></i><b>2.1</b> Setup</a></li>
<li class="chapter" data-level="2.2" data-path="quick-start.html"><a href="quick-start.html#configuration"><i class="fa fa-check"></i><b>2.2</b> Configuration</a><ul>
<li class="chapter" data-level="2.2.1" data-path="quick-start.html"><a href="quick-start.html#your-main-script"><i class="fa fa-check"></i><b>2.2.1</b> Your “main” script</a></li>
<li class="chapter" data-level="2.2.2" data-path="quick-start.html"><a href="quick-start.html#your-config-file"><i class="fa fa-check"></i><b>2.2.2</b> Your config file</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="setup-details.html"><a href="setup-details.html"><i class="fa fa-check"></i><b>3</b> Setup Details</a><ul>
<li class="chapter" data-level="3.1" data-path="setup-details.html"><a href="setup-details.html#requirements"><i class="fa fa-check"></i><b>3.1</b> Requirements</a></li>
<li class="chapter" data-level="3.2" data-path="setup-details.html"><a href="setup-details.html#installation"><i class="fa fa-check"></i><b>3.2</b> Installation</a></li>
<li class="chapter" data-level="3.3" data-path="setup-details.html"><a href="setup-details.html#run-the-pipeline"><i class="fa fa-check"></i><b>3.3</b> Run the Pipeline</a><ul>
<li class="chapter" data-level="3.3.1" data-path="setup-details.html"><a href="setup-details.html#run-in-a-slurm-environment-cluster"><i class="fa fa-check"></i><b>3.3.1</b> Run in a SLURM environment/ cluster</a></li>
<li class="chapter" data-level="3.3.2" data-path="setup-details.html"><a href="setup-details.html#run-on-a-sun-grid-engines-sge-cluster"><i class="fa fa-check"></i><b>3.3.2</b> Run on a Sun Grid Engines (SGE) cluster</a></li>
<li class="chapter" data-level="3.3.3" data-path="setup-details.html"><a href="setup-details.html#run-locally"><i class="fa fa-check"></i><b>3.3.3</b> Run locally</a></li>
<li class="chapter" data-level="3.3.4" data-path="setup-details.html"><a href="setup-details.html#run-on-the-jhpce-cluster"><i class="fa fa-check"></i><b>3.3.4</b> Run on the <span>JHPCE</span> cluster</a></li>
<li class="chapter" data-level="3.3.5" data-path="setup-details.html"><a href="setup-details.html#example-main-script"><i class="fa fa-check"></i><b>3.3.5</b> Example main script</a></li>
<li class="chapter" data-level="3.3.6" data-path="setup-details.html"><a href="setup-details.html#advanced-info-regarding-installation"><i class="fa fa-check"></i><b>3.3.6</b> Advanced info regarding installation</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="command-opts.html"><a href="command-opts.html"><i class="fa fa-check"></i><b>4</b> All Command Options</a><ul>
<li class="chapter" data-level="4.1" data-path="command-opts.html"><a href="command-opts.html#speaqeasy-options"><i class="fa fa-check"></i><b>4.1</b> SPEAQeasy Options</a><ul>
<li class="chapter" data-level="4.1.1" data-path="command-opts.html"><a href="command-opts.html#mandatory-parameters"><i class="fa fa-check"></i><b>4.1.1</b> Mandatory Parameters</a></li>
<li class="chapter" data-level="4.1.2" data-path="command-opts.html"><a href="command-opts.html#optional-parameters"><i class="fa fa-check"></i><b>4.1.2</b> Optional Parameters</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="command-opts.html"><a href="command-opts.html#nextflow-options"><i class="fa fa-check"></i><b>4.2</b> Nextflow Options</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="pipeline-overview.html"><a href="pipeline-overview.html"><i class="fa fa-check"></i><b>5</b> Pipeline Overview</a><ul>
<li class="chapter" data-level="5.1" data-path="pipeline-overview.html"><a href="pipeline-overview.html#preparation-steps"><i class="fa fa-check"></i><b>5.1</b> Preparation Steps</a><ul>
<li class="chapter" data-level="5.1.1" data-path="pipeline-overview.html"><a href="pipeline-overview.html#downloading-required-annotation-files"><i class="fa fa-check"></i><b>5.1.1</b> Downloading required annotation files</a></li>
<li class="chapter" data-level="5.1.2" data-path="pipeline-overview.html"><a href="pipeline-overview.html#preparing-annotation-files-for-direct-use"><i class="fa fa-check"></i><b>5.1.2</b> Preparing annotation files for direct use</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="pipeline-overview.html"><a href="pipeline-overview.html#main-workflow-steps"><i class="fa fa-check"></i><b>5.2</b> Main Workflow Steps</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="annotation.html"><a href="annotation.html"><i class="fa fa-check"></i><b>6</b> Annotation</a><ul>
<li class="chapter" data-level="6.1" data-path="annotation.html"><a href="annotation.html#default-annotation"><i class="fa fa-check"></i><b>6.1</b> Default Annotation</a><ul>
<li class="chapter" data-level="6.1.1" data-path="annotation.html"><a href="annotation.html#choosing-a-release"><i class="fa fa-check"></i><b>6.1.1</b> Choosing a release</a></li>
<li class="chapter" data-level="6.1.2" data-path="annotation.html"><a href="annotation.html#choosing-a-build"><i class="fa fa-check"></i><b>6.1.2</b> Choosing a “build”</a></li>
<li class="chapter" data-level="6.1.3" data-path="annotation.html"><a href="annotation.html#additional-annotation-files"><i class="fa fa-check"></i><b>6.1.3</b> Additional Annotation files</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="annotation.html"><a href="annotation.html#custom-annotation"><i class="fa fa-check"></i><b>6.2</b> Custom Annotation</a><ul>
<li class="chapter" data-level="6.2.1" data-path="annotation.html"><a href="annotation.html#optional-files-to-include-depending-on-your-use-case"><i class="fa fa-check"></i><b>6.2.1</b> Optional files to include depending on your use-case</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="manifest.html"><a href="manifest.html"><i class="fa fa-check"></i><b>7</b> Manifest and Inputs</a><ul>
<li class="chapter" data-level="7.1" data-path="manifest.html"><a href="manifest.html#what-the-manifest-should-look-like"><i class="fa fa-check"></i><b>7.1</b> What the Manifest Should Look Like</a><ul>
<li class="chapter" data-level="7.1.1" data-path="manifest.html"><a href="manifest.html#more-details-regarding-inputs"><i class="fa fa-check"></i><b>7.1.1</b> More details regarding inputs</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="outputs.html"><a href="outputs.html"><i class="fa fa-check"></i><b>8</b> Pipeline Outputs</a><ul>
<li class="chapter" data-level="8.1" data-path="outputs.html"><a href="outputs.html#main-outputs"><i class="fa fa-check"></i><b>8.1</b> Main Outputs</a></li>
<li class="chapter" data-level="8.2" data-path="outputs.html"><a href="outputs.html#intermediary-outputs"><i class="fa fa-check"></i><b>8.2</b> Intermediary Outputs</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="quick-start.html"><a href="quick-start.html#configuration"><i class="fa fa-check"></i><b>9</b> Configuration</a><ul>
<li class="chapter" data-level="9.1" data-path="configuration.html"><a href="configuration.html"><i class="fa fa-check"></i><b>9.1</b> Specifying Options for your Cluster</a><ul>
<li class="chapter" data-level="9.1.1" data-path="configuration.html"><a href="configuration.html#time"><i class="fa fa-check"></i><b>9.1.1</b> Time</a></li>
<li class="chapter" data-level="9.1.2" data-path="configuration.html"><a href="configuration.html#cluster-specific-options"><i class="fa fa-check"></i><b>9.1.2</b> Cluster-specific options</a></li>
</ul></li>
<li class="chapter" data-level="9.2" data-path="configuration.html"><a href="configuration.html#speaqeasy-parameters"><i class="fa fa-check"></i><b>9.2</b> SPEAQeasy Parameters</a><ul>
<li class="chapter" data-level="9.2.1" data-path="configuration.html"><a href="configuration.html#annotation-settings"><i class="fa fa-check"></i><b>9.2.1</b> Annotation settings</a></li>
<li class="chapter" data-level="9.2.2" data-path="configuration.html"><a href="configuration.html#software-settings"><i class="fa fa-check"></i><b>9.2.2</b> Software settings</a></li>
<li class="chapter" data-level="9.2.3" data-path="configuration.html"><a href="configuration.html#miscellaneous-settings"><i class="fa fa-check"></i><b>9.2.3</b> Miscellaneous settings</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="10" data-path="software.html"><a href="software.html"><i class="fa fa-check"></i><b>10</b> Software Versions</a><ul>
<li class="chapter" data-level="10.1" data-path="software.html"><a href="software.html#docker-images"><i class="fa fa-check"></i><b>10.1</b> Docker Images</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="help.html"><a href="help.html"><i class="fa fa-check"></i><b>11</b> Help</a><ul>
<li class="chapter" data-level="11.1" data-path="help.html"><a href="help.html#common-errors"><i class="fa fa-check"></i><b>11.1</b> Common Errors</a></li>
<li class="chapter" data-level="11.2" data-path="help.html"><a href="help.html#deeper-investigation"><i class="fa fa-check"></i><b>11.2</b> Deeper Investigation</a><ul>
<li class="chapter" data-level="11.2.1" data-path="help.html"><a href="help.html#checking-per-sample-logs"><i class="fa fa-check"></i><b>11.2.1</b> Checking Per-Sample Logs</a></li>
</ul></li>
<li class="chapter" data-level="11.3" data-path="help.html"><a href="help.html#resuming"><i class="fa fa-check"></i><b>11.3</b> Resuming</a><ul>
<li class="chapter" data-level="11.3.1" data-path="help.html"><a href="help.html#trouble-resuming"><i class="fa fa-check"></i><b>11.3.1</b> Trouble resuming?</a></li>
</ul></li>
<li class="chapter" data-level="11.4" data-path="help.html"><a href="help.html#no-internet-access"><i class="fa fa-check"></i><b>11.4</b> No Internet Access?</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="glossary.html"><a href="glossary.html"><i class="fa fa-check"></i><b>12</b> Glossary</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">SPEAQeasy</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="outputs" class="section level1">
<h1><span class="header-section-number">8</span> Pipeline Outputs</h1>
<div id="main-outputs" class="section level2">
<h2><span class="header-section-number">8.1</span> Main Outputs</h2>
<div class="figure">
<img src="images/7_main_outputs.png" alt="" />
<p class="caption">Main Outputs</p>
</div>
<p>Most importantly, SPEAQeasy generates <a href="https://bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html"><code>RangedSummarizedExperiment</code></a> objects storing counts at several features (genes, exons, and exon-exon junctions). A number of <a href="https://bioconductor.org/">Bioconductor</a> packages can be trivially utilized to perform desired differential expression analyses from counts in the widespread <code>SummarizedExperiment</code> format, and we provide a <a href="NEEDS%20A%20LINK!">vignette</a> demonstrating an example analysis.</p>
<p>For human samples, variants are called at a list of common missense single nucleotide variants (SNVs), and SPEAQeasy ultimately a single VCF file to store genotype calls at these sites for all samples in the experiment. We provide a <a href="NEEDS%20A%20LINK!">guide</a> walking through how this genotype data can be used to resolve identity issues that arise during sequencing, salvaging samples which otherwise might be dropped from further analysis.</p>
<p>Finally, expressed regions data is optionally generated (with the <code>--coverage</code> <a href="command-opts.html#command-opts">option</a>). The <code>RData</code> files provide a starting point for finding differentially expressed regions (DERs), for analyses involving this end goal.</p>
</div>
<div id="intermediary-outputs" class="section level2">
<h2><span class="header-section-number">8.2</span> Intermediary Outputs</h2>
<p>SPEAQeasy generates a number of files along the process before producing the main outputs of interest. Each of these “intermediary” files is described below.</p>
<table>
<colgroup>
<col width="19%" />
<col width="20%" />
<col width="59%" />
</colgroup>
<thead>
<tr class="header">
<th>Filename</th>
<th>Directory</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>ercc_spikein_check_mix1.pdf</td>
<td>count_objects/</td>
<td>?</td>
</tr>
<tr class="even">
<td>rawCounts_[experiment_name]_n[num_samples].rda</td>
<td>count_objects/</td>
<td>An R data file containing a number of matrices/ data frames containing raw gene, exon, exon-exon junction, and transcript counts. A data frame of quality metrics is also included.</td>
</tr>
<tr class="odd">
<td>read_and_alignment_metrics_[experiment_name].csv</td>
<td>count_objects/</td>
<td>A comma-separated values file of quality metrics.</td>
</tr>
<tr class="even">
<td>rpkm_counts_[experiment_name]_n[num_samples].rda</td>
<td>count_objects/</td>
<td>The same data as in rawCounts_[experiment_name]_n[num_samples].rda, but normalized as reads-per-kilobase-million (RPKM).</td>
</tr>
<tr class="odd">
<td>[sample_name]_[annotation_version]_Exons.counts</td>
<td>counts/</td>
<td>Exon counts for each sample individually, as reported by <a href="http://bioinf.wehi.edu.au/featureCounts/">featureCounts</a></td>
</tr>
<tr class="even">
<td>[sample_name]_[annotation_version]_Exons.counts.summary</td>
<td>counts/</td>
<td>A summary of key information generated from quantifying exons on a particular sample with <a href="http://bioinf.wehi.edu.au/featureCounts/">featureCounts</a></td>
</tr>
<tr class="odd">
<td>[sample_name]_[annotation_version]_Gene.counts</td>
<td>counts/</td>
<td>Gene counts for each sample individually, as reported by <a href="http://bioinf.wehi.edu.au/featureCounts/">featureCounts</a></td>
</tr>
<tr class="even">
<td>[sample_name]_[annotation_version]_Gene.counts.summary</td>
<td>counts/</td>
<td>A summary of key information generated from quantifying genes on a particular sample with <a href="http://bioinf.wehi.edu.au/featureCounts/">featureCounts</a></td>
</tr>
<tr class="odd">
<td>[sample_name]_junctions_primaryOnly_regtools.bed</td>
<td>counts/junction/</td>
<td>A file in BED format describing junctions determined by <a href="https://regtools.readthedocs.io/en/latest/">regtools</a> run on primary alignments for a given sample</td>
</tr>
<tr class="even">
<td>[sample_name]_junctions_primaryOnly_regtools.count</td>
<td>counts/junction/</td>
<td>A text file describing the junction ranges and providing raw counts of hits for each junction, as output by <a href="https://regtools.readthedocs.io/en/latest/">regtools</a></td>
</tr>
<tr class="odd">
<td>[sample_name].bam and [sample_name].bam.bai</td>
<td>counts/junction/primary_alignments/</td>
<td>The subset of primary alignments for a given sample, in BAM format, along with an index for each BAM.</td>
</tr>
<tr class="even">
<td>mean.forward.bw and mean.reverse.bw</td>
<td>coverage/mean/</td>
<td>BigWig files containing coverage information averaged across all samples in the experiment, and split by strand.</td>
</tr>
<tr class="odd">
<td>[sample_name].[strand].wig</td>
<td>coverage/wigs/</td>
<td>Wiggle files containing coverage information for a given sample and strand.</td>
</tr>
<tr class="even">
<td>abundance.tsv</td>
<td>ERCC/[sample_name]/</td>
<td>[TODO]</td>
</tr>
<tr class="odd">
<td>region_cuts_[strand].Rdata</td>
<td>expressed_regions/</td>
<td>[TODO]</td>
</tr>
<tr class="even">
<td>region_cuts_raw_[strand].Rdata</td>
<td>expressed_regions/</td>
<td>[TODO]</td>
</tr>
<tr class="odd">
<td>region_info_[strand].Rdata</td>
<td>expressed_regions/</td>
<td>[TODO]</td>
</tr>
<tr class="even">
<td>region_info_[strand].pdf</td>
<td>expressed_regions/</td>
<td>[TODO]</td>
</tr>
<tr class="odd">
<td>[trim_status]/[file_name]/*</td>
<td>fastQC/</td>
<td>Outputs from <a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a>. Here trim_status indicates when FastQC was performed: Untrimmed is before trimming, and Trimmed is after. file_name contains the sample name, and if applicable, the mate number.</td>
</tr>
<tr class="even">
<td>[sample_name]_hisat_out.sam</td>
<td>HISAT2_out/</td>
<td>The main alignment output from <a href="http://daehwankimlab.github.io/hisat2/">Hisat2</a> in SAM format</td>
</tr>
<tr class="odd">
<td>[sample_name]_align_summary.txt</td>
<td>HISAT2_out/</td>
<td>The text-based alignment summary from <a href="http://daehwankimlab.github.io/hisat2/">Hisat2</a>. Note that metrics from these files are aggregated for the experiment, and so users likely will not need to check or process the original files manually (metrics).</td>
</tr>
<tr class="even">
<td>[sample_name]_accepted_hits.sorted.bam and [sample_name]_accepted_hits.sorted.bam.bai</td>
<td>HISAT2_out/sam_to_bam/</td>
<td>Alignments from <a href="http://daehwankimlab.github.io/hisat2/">Hisat2</a> where unmapped reads are filtered out, and the resulting BAM file is coordinate-sorted and indexed.</td>
</tr>
<tr class="odd">
<td>[sample_name]_strandness_pattern.txt</td>
<td>infer_strandness/</td>
<td>A text file containing the inferred strandness pattern for each sample. These files are primarily for internal use by the pipeline, and it is recommended to check the output file <code>samples_complete.manifest</code> to quickly view strandness patterns for all samples.</td>
</tr>
<tr class="even">
<td>samples_complete.manifest</td>
<td>infer_strandness/</td>
<td>A version of the input samples.manifest, with an additional column listing the inferred strandness pattern for each sample</td>
</tr>
<tr class="odd">
<td>*</td>
<td>kallisto_tx/[sample_name]/</td>
<td>[TODO]</td>
</tr>
<tr class="even">
<td>[sample_name]_process_trace.log</td>
<td>logs/</td>
<td>A SPEAQeasy-generated log tracing the processes and associated commands run for each sample. This is intended to help users quickly determine the source of any errors during pipeline execution (text).</td>
</tr>
<tr class="odd">
<td>[sample_name]_trimmed*.fastq</td>
<td>trimming/</td>
<td>Trimmed FASTQ files, if applicable, from <a href="http://www.usadellab.org/cms/?page=trimmomatic">Trimmomatic</a></td>
</tr>
</tbody>
</table>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="manifest.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="configuration.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
