# LIEBER INSTITUTE JAFFE LAB RNA - SEQ ANALYSIS PIPELINE #

### Summary ###
_##TODO##_: Brief description of the pipeline

_##TODO##_: Show a simplified workflow of the pipeline

### Version description ###

* Version 0.7.3 (current)

 + _##TODO##_: put in the features up to this version

### Installation ###

##### OS #####

_##TODO##_: describe in what OS this has been tested

##### Bioinformatic software #####

_##TODO##_: Finish software and versions, according to what is installed in WG test server. a lot of software is missing

* [FastQC](http://www.bioinformatics.babraham.ac.uk/projects/fastqc)
* [Trimmomatic](http://www.usadellab.org/cms/?page=trimmomatic)
* [BWA](http://bio-bwa.sourceforge.net/)
* [Picard tools](http://broadinstitute.github.io/picard/)
* [GATK](https://www.broadinstitute.org/gatk/)
* [muTect](https://www.broadinstitute.org/cancer/cga/mutect)
* [Pindel](http://gmt.genome.wustl.edu/packages/pindel/)
* [samtools](http://samtools.sourceforge.net/)
* [tabix](http://www.htslib.org/doc/tabix.html)
* [vcftools](http://vcftools.sourceforge.net/)
* [bedtools](http://bedtools.readthedocs.org/en/latest/)
* [snpEff](http://snpeff.sourceforge.net/)
* [mk](https://github.com/dcjones/mk)

### Configuration ###

* _##TODO##_: describe nextflow.config file

* _##TODO##_: describe conf/command_paths.config

* _##TODO##_: describe conf/docker.config

* _##TODO##_: describe conf/mem.config

* _##TODO##_: describe conf/sge.config
* _##TODO##_: describe conf/sge_large.config


### Reference files ###

_##TODO##_: Briefly describe Anno and Genotyping directories

_##TODO##_: Show a tree view of the directories, showing which files are versioned and which are built by the pipeline

### Test run ###

_##TODO##_: describe how to launch a test run (via bash run_test.sh)

### Process description ###

_##TODO##_: Describe for every process step of the pipeline, what it does (order it accordingly to what it is described in the main.nf header)

### Input data formats ###

_##TODO##_: describe species and data types (single, paired, etc.) accepted by the pipeline

### Output data formats ###

_##TODO##_: describe how to install NF

### Launching a real run ###

_##TODO##_: describe how to install NF

### Run with Docker ###

_##TODO##_: describe how to install NF

### Run with SGE ###

_##TODO##_: describe how to install NF

### Pipeline directory structure ###

_##TODO##_: describe how to install NF

### Contact ###

_##TODO##_: describe how to install NF

* [Leonardo Collado Torres](http://lcolladotor.github.io/)
* [Winter Genomics Team](http://www.wintergenomics.com)
