# LIEBER INSTITUTE JAFFE LAB RNA - SEQ ANALYSIS PIPELINE #

### Summary ###
_##TODO##_: Brief description of the pipeline

_##TODO##_: Show a simplified workflow of the pipeline, from a notes/*.png image (done in draw.io or something similar)

### Version description ###

* Version 0.7.3 (current)

 + _##TODO##_: put in the features up to this version

### Installation ###

##### OS #####

_##TODO##_: describe in what OS this has been tested

##### Bioinformatic software #####

_##TODO##_: Finish software and versions, according to what is installed in WG test server. a lot of software is missing

* [FastQC](http://www.bioinformatics.babraham.ac.uk/projects/fastqc)

### Configuration ###

* _##TODO##_: describe nextflow.config file

* _##TODO##_: describe conf/command_paths.config

* _##TODO##_: describe conf/docker.config

* _##TODO##_: describe conf/mem.config

* _##TODO##_: describe conf/sge.config
* _##TODO##_: describe conf/sge_large.config


### Reference files ###

_##TODO##_: Briefly describe Anno and Genotyping directories

_##TODO##_: Show a tree view of the directories, showing which files are versioned and which are built by the pipeline

### Test run ###

_##TODO##_: describe how to launch a test run (via bash run_test.sh)

### Process description ###

_##TODO##_: Describe for every process step of the pipeline, what it does (order it accordingly to what it is described in the main.nf header)

### Input data formats ###

_##TODO##_: describe species and data types (single, paired, etc.) accepted by the pipeline.

_##TODO##_: Make notes about file naming, for normal runs, and for --merged runs

_##TODO##_: Make notes about disabled modules for mm10 and rn6, if any

### Output data formats ###

_##TODO##_: describe the many output files produced by the pipeline.

_##TODO##_: Consult with Lieber which files are final outputs and which are temporary files

### Launching a real run ###

_##TODO##_: describe how to launch a normal run

_##TODO##_: describe the many options of the pipeline, and the flags, and what they mean

### Run with Docker ###

_##TODO##_: describe how docker integration works in this pipeline

_##TODO##_: describe how to build or pull the dockers.

_##TODO##_: describe the --with-docker flag

### Run with SGE ###

_##TODO##_: describe how SGE integration works in this pipeline

_##TODO##_: describe how to configure the cong/sge* configuration files, regarding queue, profile and resources request

### Pipeline directory structure ###

_##TODO##_: make a tree view of a final directory print from all the sucessfull runs.

_##TODO##_: describe in brief wvery file in the tree

### Contact ###

_##TODO##_: add author list

* [Leonardo Collado Torres](http://lcolladotor.github.io/)
* [Winter Genomics Team](http://www.wintergenomics.com)
